### PR TITLE
[java] New Rule - UseUnderscoresInNumericLiterals

### LINKED ISSUE #1232: [java] Detector for large numbers not separated by _
Java allows numbers like `1000000` to be written like `1_000_000` which is much more readable. A detector that warns about numbers that have more than 3 digits not separated by `_` would be nice.

### PR DESCRIPTION
**Pull request by rajeshggwp**
Before submitting a PR, please check that:
 - [x] The PR is submitted against `master`. The PMD team will merge back to support branches as needed.
 - [x] `./mvnw clean verify` passes. This will [build](https://github.com/pmd/pmd/blob/master/BUILDING.md) and test PMD, execute PMD and checkstyle rules. [Check this for more info](https://github.com/pmd/pmd/blob/master/CONTRIBUTING.md#code-style)

**PR Description:**
Resolves #1232

### CODE REVIEW

**Comment by jsotuyod** [comment]
@rajeshggwp thanks for the PR!

The rule seems to be introducing false positives, for instance, it's reporting on this line:

https://github.com/checkstyle/checkstyle/blob/9cf091d553be0567e68e0439f56f12c63c496e18/src/it/java/com/google/checkstyle/test/base/AbstractIndentationTestSupport.java#L41

I'd add a new test case reproducing the issue, and then work on fixing it. Pushing to your branch will have the regression tester rerun and update results here automatically :)

**Comment by jsotuyod** [comment]
The rule is working much better now! Looking at the pmd-test results I see however some scenarios we could handle better:

1. `serialVersionUID` is autogenerated and developers don't really care about reading it. We should probably not report on that field
2. numbers that are not in base 10 shouldn't be reported (if not using base 10, the 3 digit rule is pointless). You can discard literals starting with '0' (covers all octal, hexa and binary)
3. your regexp is not considering scientific notation. `-2.147483648E10` is valid.
4. your regexp is not allowing underscores in the decimal part. I can understand you not wanting to enforce them, but I'm unsure disallowing them altogether is sane. In the precious example, I'd personally think `-2.147_483_648E10` is clearer to read.

Finally, even though your rule handles it properly, I'd like to include a couple tests with negative numbers (even negatives in bases other than 10, as per the second item). It should just work as the rule stands, but keeping it safe with a test.

File: pmd-java/src/main/resources/category/java/codestyle.xml [hunk 1]
```diff
@@ -1351,6 +1351,40 @@ public class ClassInDefaultPackage {
         </example>
     </rule>
 
+    <rule name="NumericLiteralConvention"
```
Comment (oowekyala) [inline pmd-java/src/main/resources/category/java/codestyle.xml]: I think naming the rule "UseUnderscoresInNumericLiterals" may be more adapted, though it's open to discussion. Unless we want to add more power to the rule, e.g. enforce using always upper- or lowercase for the `d` or `f` suffix

File: pmd-java/src/main/resources/category/java/codestyle.xml [hunk 2]
```diff
@@ -1351,6 +1351,40 @@ public class ClassInDefaultPackage {
...
+        <properties>
+            <property name="version" value="2.0"/>
+            <property name="xpath">
+                <value>
+                    <![CDATA[
+//Literal[@IntLiteral = true() or
+ @LongLiteral = true() or
+ @DoubleLiteral = true() or
+ @FloatLiteral = true()]
+ [not(matches(@Image, "^[0-9]{1,3}(_[0-9]{3})*(l|L|\.[0-9]+)?(d|D|f|F)?$"))]
```
Comment (oowekyala) [inline pmd-java/src/main/resources/category/java/codestyle.xml]: One alternation may be simplified:

```suggestion
 [not(matches(@Image, "^[0-9]{1,3}(_[0-9]{3})*(l|L|\.[0-9]+)?[dDfF]?$"))]
```

There should probably be a configurable minimum length threshold before the underscores are required. For instance for 4 digit numbers: I don't see a problem reading `1852`, whereas `1_852` is slightly less readable IMO.

File: pmd-java/src/main/resources/category/java/codestyle.xml (hunk 1 above)
Comment (rajeshggwp) [reply pmd-java/src/main/resources/category/java/codestyle.xml]: This certainly makes sense. I think for now we can go ahead with "UseUnderscoresInNumericLiterals" and later this can be renamed if more rules related to numeric literal convention is added.

File: pmd-java/src/main/resources/category/java/codestyle.xml (hunk 2 above)
Comment (rajeshggwp) [reply pmd-java/src/main/resources/category/java/codestyle.xml]: Could you let me know how configuring the length can be achieved?

Comment (oowekyala) [reply pmd-java/src/main/resources/category/java/codestyle.xml]: @rajeshggwp Yes, using a property is the way to go. We could split the string before and after the decimal separator, removing the `dfl` suffixes. Then each segment is *valid* iff either 
* it is shorter than the acceptable length, and either doesn't have any underscores, or all its underscores are well placed
* it is longer than the acceptable length, and all its underscores are well placed

If either segment is invalid, then we report the node. The following condition does the trick:
```ruby
[
  some $num in tokenize(@Image, "[.dDfFlL]") 
  satisfies not(
                 string-length($num) <= $acceptableLength 
                   and (
                         not(contains($num,"_")) 
                         or matches($num, "^[0-9]{1,3}(_[0-9]{3})*$")
                       )
		 or matches($num, "^[0-9]{1,3}(_[0-9]{3})*$")
               )
 ]
```

File: pmd-java/src/main/resources/category/java/codestyle.xml [hunk 3]
```diff
@@ -1351,6 +1351,42 @@ public class ClassInDefaultPackage {
...
+        <properties>
+            <property name="version" value="2.0"/>
+            <property name="xpath">
+                <value>
+                    <![CDATA[
+//Literal[@IntLiteral = true() or
+ @LongLiteral = true() or
+ @DoubleLiteral = true() or
+ @FloatLiteral = true()]
+ [not (matches(@Image, "^(0([xb]?[0-9a-fA-F]+)?|[1-9][0-9]{0,2}(_[0-9]{3})*)(l|L|\.[0-9_]+)?([eE][\+-]?[0-9]+)?[dDfF]?$"))]
```
Comment (oowekyala) [inline pmd-java/src/main/resources/category/java/codestyle.xml]: I think instead of ignoring bases other than decimal, it would be nice to require them to be underscored too.
E.g. hex literals are sometimes divided in groups of two digits, bc two hex digits make a byte. 

But then, the maximum acceptable length would need to be configured base by base.

Better keep this PR simple and implement that later. I'd prefer the acceptable length property for decimal literals to be added in this PR though.

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/UseUnderscoresInNumericLiterals.xml [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+        ]]></code>
+    </test-code>
+    <test-code>
+        <description><![CDATA[
+ok, Numeric Literal with negative exponent
+     ]]></description>
+        <expected-problems>0</expected-problems>
+        <code><![CDATA[
+public class Foo {
+    int binary = 36e-11234;
```
Comment (oowekyala) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/UseUnderscoresInNumericLiterals.xml]: Shouldn't we recommend the exponent to be written as `11_234`?

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/UseUnderscoresInNumericLiterals.xml [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+}
+        ]]></code>
+    </test-code>
+    <test-code>
+        <description><![CDATA[
```
Comment (oowekyala) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/UseUnderscoresInNumericLiterals.xml]: Please don't wrap the description of tests in a CDATA block. They nearly never require it and it makes the test file harder to read.

File: pmd-java/src/main/resources/category/java/codestyle.xml (hunk 3 above)
Comment (rajeshggwp) [reply pmd-java/src/main/resources/category/java/codestyle.xml]: Since, the bases other than decimal follow a different grouping strategy and 3 digit grouping does not work well with these bases, it has been ignored. We can extend this rule to support other bases later on.
I will implement the acceptable length property for decimal literals in this PR.

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/UseUnderscoresInNumericLiterals.xml (hunk 4 above)
Comment (rajeshggwp) [reply pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/UseUnderscoresInNumericLiterals.xml]: Oops! 
Yes, you are right and I am working on this.

File: pmd-java/src/main/resources/category/java/codestyle.xml [hunk 6]
```diff
@@ -1366,14 +1366,27 @@ public class ClassInDefaultPackage {
         <priority>3</priority>
         <properties>
             <property name="version" value="2.0"/>
+            <property name="acceptableLength" type="Integer" value="4"
```
Comment (oowekyala) [inline pmd-java/src/main/resources/category/java/codestyle.xml]: You need to add a `min` & `max` attribute for it to work

Comment (oowekyala) [reply pmd-java/src/main/resources/category/java/codestyle.xml]: Also, since we plan on extending the rule to other bases, naming it "acceptableDecimalLength" would make it future-proof since we'd probably add e.g. "acceptableHexLength"

File: pmd-java/src/main/resources/category/java/codestyle.xml [hunk 7]
```diff
@@ -1366,14 +1366,27 @@ public class ClassInDefaultPackage {
         <priority>3</priority>
         <properties>
             <property name="version" value="2.0"/>
+            <property name="acceptableDecimalLength" type="Integer" value="4" min="3" max="6"
```
Comment (oowekyala) [inline pmd-java/src/main/resources/category/java/codestyle.xml]: Placing the max at 6 seems arbitrary. I think it's not PMD's problem if users want to place it at 10. The rule would be less useful, but it also checks for misplaced underscores if any, so, whatever 

Ideally it would only be required to be positive, but we need to provide a max anyway (refs #1204), which will be fixed in 7.0.0 (#1432). I think we'd better give it a very high value so that it doesn't get in the way for now

File: pmd-java/src/main/resources/category/java/codestyle.xml [hunk 8]
```diff
@@ -1353,6 +1353,55 @@ public class ClassInDefaultPackage {
         </example>
     </rule>
 
+    <rule name="UseUnderscoresInNumericLiterals"
+          language="java"
+          since="6.10.0"
+          minimumLanguageVersion="1.7"
+          message="Numeric literals with more than 3 digits must use '_' as a separator"
```
Comment (oowekyala) [inline pmd-java/src/main/resources/category/java/codestyle.xml]: Maybe the rule message can be improved by using the placeholder `{0}`, which will be replaced by the image of the matched node, i.e. by the number. E.g. `Number {0} should separate every third digit with an underscore`. This message also passes when the violation is because of misplaced underscores

File: pmd-java/src/main/resources/category/java/codestyle.xml [hunk 9]
```diff
@@ -1353,6 +1353,55 @@ public class ClassInDefaultPackage {
...
+ @DoubleLiteral = true() or
+ @FloatLiteral = true()]
+ [ not (matches(@Image, "^0([xb]?[0-9a-fA-F]+)?$"))]
+ [
+  some $num in tokenize(@Image, "[.dDfFlL]|[eE][\+-]?")
+  satisfies not(
+                 string-length($num) <= $acceptableDecimalLength
+                   and (
+                         not(contains($num,"_"))
+                         or matches($num, "^[0-9]{1,3}(_[0-9]{3})*$")
```
Comment (jsotuyod) [inline pmd-java/src/main/resources/category/java/codestyle.xml]: this line can probably be dropped...
checking "if under max length and has no underscore OR matches the regexp" should already cover all scenarios without duplicating the regexp

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/UseUnderscoresInNumericLiterals.xml [hunk 10]
```diff
@@ -0,0 +1,306 @@
...
+        ]]></code>
+    </test-code>
+    <test-code>
+        <description>
+            ok, Lengthy numeric literal with variable name as serialVersionUID
+        </description>
+        <expected-problems>0</expected-problems>
+        <code><![CDATA[
+public class Foo {
+    int serialVersionUID = -2518308549741147689L;
```
Comment (jsotuyod) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/UseUnderscoresInNumericLiterals.xml]: try using an actual `serialVersionUID` (`static final long`) just to be on the safe side

**Comment by rajeshggwp** [comment]
why am I not able to see travis build details? is the result purged already?

File: pmd-java/src/main/resources/category/java/codestyle.xml (hunk 8 above)
Comment (rajeshggwp) [reply pmd-java/src/main/resources/category/java/codestyle.xml]: This works fine! Could you tell me how {0} is associated to Image?

Comment (oowekyala) [reply pmd-java/src/main/resources/category/java/codestyle.xml]: Java rules may report a violation with an array of arguments, which are interpolated index-by-index on placeholder like `{0}`, `{1}`, etc. As it happens XPath rules have no flexible way to add arguments to a message, but the image of the matched node is always passed nevertheless, so is always accessible as `{0}` (for XPath rules only). See here:

https://github.com/pmd/pmd/blob/0c1afc426debff979b25219dbc8f4e07c76645b5/pmd-core/src/main/java/net/sourceforge/pmd/lang/rule/XPathRule.java#L114

**Comment by jsotuyod** [comment]
This is awesome! It seems ready to merge. I'll do so when I get the chance.

File: pmd-java/src/main/resources/category/java/codestyle.xml [hunk 11]
```diff
@@ -1353,6 +1353,54 @@ public class ClassInDefaultPackage {
...
+            <property name="acceptableDecimalLength" type="Integer" value="4" min="3" max="1000"
+                      description="Maximum acceptable length within which the rule is not applicable"/>
+            <property name="xpath">
+                <value>
+                    <![CDATA[
+//Literal[@IntLiteral = true() or
+ @LongLiteral = true() or
+ @DoubleLiteral = true() or
+ @FloatLiteral = true()]
+ [ not (matches(@Image, "^0([xb]?[0-9a-fA-F]+)?$"))]
```
Comment (oowekyala) [inline pmd-java/src/main/resources/category/java/codestyle.xml]: I think this may be simplified as `^0[xb]?.*`

File: pmd-java/src/main/resources/category/java/codestyle.xml [hunk 12]
```diff
@@ -1353,6 +1353,54 @@ public class ClassInDefaultPackage {
...
+ [
+  some $num in tokenize(@Image, "[.dDfFlL]|[eE][\+-]?")
```
Comment (oowekyala) [inline pmd-java/src/main/resources/category/java/codestyle.xml]: This could be simplified as `[.dDfFlLeE+-]`. Empty strings in the sequence are discarded immediately by the string-length condition.

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/UseUnderscoresInNumericLiterals.xml [hunk 13]
```diff
@@ -0,0 +1,318 @@
...
+    </test-code>
+    <test-code>
+        <description>
+            bad, testing acceptableDecimalLength
+        </description>
+        <rule-property name="acceptableDecimalLength">10</rule-property>
+        <expected-problems>1</expected-problems>
+        <code><![CDATA[
+public class Foo {
+    int value = -32_3639_6_9;
```
Comment (oowekyala) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/UseUnderscoresInNumericLiterals.xml]: This test should test a number that is under the acceptableDecimalLength to check that misplaced underscores are flagged. This one is 11 chars long ^^

Comment (rajeshggwp) [reply pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/UseUnderscoresInNumericLiterals.xml]: are you thinking of a test case similar to this one - https://github.com/pmd/pmd/pull/1384/files/9b6b1a4d15b7013b302467edab093f941f5d8fc4#diff-23912dc999320f5a64c4d06d86bdfdd2R96 ?
This one does not have acceptableDecimalLength defined, so it gets defaulted to 4.

File: pmd-java/src/main/resources/category/java/codestyle.xml (hunk 11 above)
Comment (rajeshggwp) [reply pmd-java/src/main/resources/category/java/codestyle.xml]: I thought of this initially but then it does not hold good with values like 0.12345.
It does not catch issues with values - 0.[0-9]+

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/UseUnderscoresInNumericLiterals.xml (hunk 13 above)
Comment (oowekyala) [reply pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/UseUnderscoresInNumericLiterals.xml]: In fact the test case you linked tests exactly the same thing as this one, since the number is longer than the acceptable length and the underscores are misplaced in both cases. I'd like a test checking that misplaced underscores are flagged even if the length is acceptable. eg `acceptableDecimalLength=10` and test on `3639_639` (length 8)

File: pmd-java/src/main/resources/category/java/codestyle.xml (hunk 11 above)
Comment (oowekyala) [reply pmd-java/src/main/resources/category/java/codestyle.xml]: Oh ok! I think maybe we may be as bold as using `^0[^.]` then?

Comment (rajeshggwp) [reply pmd-java/src/main/resources/category/java/codestyle.xml]: yes, this does the trick!

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/UseUnderscoresInNumericLiterals.xml (hunk 13 above)
Comment (rajeshggwp) [reply pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/UseUnderscoresInNumericLiterals.xml]: my bad! I got confused. Modified the length now.
